### PR TITLE
chore(notifications): unit tests and refactor to include Android utils changes 

### PR DESCRIPTION
**Pull request by Samaritan1011001**
*Issue #, if available:*

*Description of changes:*
Contains the following major changes,
- Updated NotificationPayload structure based on the changes from released shared Android utils 
- Unit tests for the Dart code
- Unit tests for Android native code

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_core/lib/src/types/notifications/push/push_notification_message.dart [hunk 1]
```diff
@@ -30,42 +30,50 @@ class PushNotificationMessage
...
-            : DateTime.fromMillisecondsSinceEpoch(sentTimeInt);
-        fcmOptions = FcmPlatformOptions(
-          channelId: fcmOptionsMap['channelId'] as String?,
-          messageId: fcmOptionsMap['messageId'] as String?,
-          senderId: fcmOptionsMap['sender'] as String?,
-          sentTime: sentTime,
-        );
       }
+    } on Error catch (e) {
+      safePrint('Error when converting Json to PushNotificationMessage: $e');
```
Comment (HuiSF) [inline packages/amplify_core/lib/src/types/notifications/push/push_notification_message.dart:76]: What's the source of error here if the code has already handled fields nullability correctly?

File: packages/notifications/push/amplify_push_notifications_pinpoint/lib/src/pinpoint_provider.dart [hunk 2]
```diff
@@ -72,18 +80,21 @@ class PinpointProvider implements ServiceProviderClient {
   }) async {
     try {
       if (!_isInitialized) {
-        _logger.error(
-          'Pinpoint provider not configured, re-run Amplify.configure',
+        throw Exception(
+          'Pinpoint provider is not initialized. Re-run Amplify.configure.',
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications_pinpoint/lib/src/pinpoint_provider.dart:84]: This can't be generic `Exception` IMO
We need proper error message, recovery suggestion and underlying exception (if any).

In addition "re-run" is not accurate, Amplify doesn't support re-configure, so when the exception happens, the only case is that configure has never been called.

File: packages/notifications/push/amplify_push_notifications_pinpoint/lib/src/push_notifications_background_processing.dart [hunk 3]
```diff
@@ -19,14 +19,17 @@ const _backgroundChannel = MethodChannel(
...
+  AmplifySecureStorage? amplifySecureStorage,
+}) async {
   WidgetsFlutterBinding.ensureInitialized();
-  final amplifySecureStorage = AmplifySecureStorage(
-    config: AmplifySecureStorageConfig(
-      scope: 'amplifyPushNotifications',
-    ),
-  );
-  final amplifyconfigStr = await amplifySecureStorage.read(
+  final amplifySecureStorage0 = amplifySecureStorage ??
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications_pinpoint/lib/src/push_notifications_background_processing.dart:26]: Setting the default value in the parameter list, so `amplifySecureStorage` can be non-nullable.

File: packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart [hunk 4]
```diff
@@ -75,21 +106,29 @@ class AmplifyPushNotifications extends PushNotificationsPluginInterface {
...
-
   late final Stream<PushNotificationMessage> _onNotificationOpened;
-
   var _isConfigured = false;
   PushNotificationMessage? _launchNotification;
   final Future<void> Function() _backgroundProcessor;
+
+  final bool _androidCheck =
+      Zone.current[androidCheckOverride] as bool? ?? Platform.isAndroid;
+  final bool _iosCheck =
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart:124]:
```suggestion
  final bool _isIOS =
```

File: packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart [hunk 5]
```diff
@@ -29,12 +31,29 @@ const _foregroundNotificationEventChannel = EventChannel(
...
 
-const _externalHandleKey = 'externalHandleKey';
+/// {@template amplify_push_notifications.external_handle_key}
+/// Key used to store the external callback handle function.
+///
+/// Used in this file to store the callback function and used in Flutter API class to access it.
+/// {@endtemplate}
+const externalHandleKey = 'externalHandleKey';
 
 const _needsConfigurationException = PushNotificationException(
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart:41]: Make this a `ConfigurationError`

File: packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart [hunk 6]
```diff
@@ -29,12 +31,29 @@ const _foregroundNotificationEventChannel = EventChannel(
...
   'Configure Amplify with Notifications Plugin before using this method.',
 );
 
+/// {@template amplify_push_notifications.android_check_override}
+/// Internal Platform check exposed for testing purposes only.
+/// {@endtemplate}
+@visibleForTesting
+const androidCheckOverride = #_androidCheckOverride;
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart:49]: This basically reimplements the functionality of `package:os_detect`. I would probably recommend moving to that. You can see example usage in `aws_common`.

File: packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+final AmplifyLogger _logger = AmplifyLogger.category(Category.pushNotifications)
+    .createChild('AmplifyPushNotification');
+
+/// {@template amplify_push_notifications.amplify_push_notifications_flutter_api}
+/// Internal Platform check exposed for testing purposes only.
+/// {@endtemplate}
+class AmplifyPushNotificationsFlutterApi
+    implements PushNotificationsFlutterApi {
+  /// {@macro amplify_push_notifications.amplify_push_notifications_flutter_api}
+  AmplifyPushNotificationsFlutterApi() {
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart:19]: Make this a singleton

File: packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    PushNotificationsFlutterApi.setup(this);
+  }
+
+  final _eventQueue = <PushNotificationMessage>[];
+
+  /// {@template amplify_push_notifications.event_queue}
+  /// Internal eventQueue getter exposed for testing purposes only.
+  /// {@endtemplate}
+  @visibleForTesting
+  List<PushNotificationMessage> get eventQueue => _eventQueue;
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart:29]:
```suggestion
  List<PushNotificationMessage> get eventQueue => UnmodifiableListView(_eventQueue);
```

File: packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  final _onNotificationReceivedInBackgroundCallbacks =
+      <OnRemoteMessageCallback>[];
+
+  /// {@template amplify_push_notifications.on_notification_received_in_background_callbacks}
+  /// Internal callbacks list getter exposed for testing purposes only.
+  /// {@endtemplate}
+  @visibleForTesting
+  List<OnRemoteMessageCallback>
+      get onNotificationReceivedInBackgroundCallbacks =>
+          _onNotificationReceivedInBackgroundCallbacks;
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart:40]:
```suggestion
          UnmodifiableListView(_onNotificationReceivedInBackgroundCallbacks);
```

File: packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        'Could not locate external callback for the stored external handle',
+      );
+      return;
+    }
+    if (externalCallback is! OnRemoteMessageCallback) {
+      throw PushNotificationException(
+        'Invalid callback type: ${externalCallback.runtimeType}',
+      );
+    }
+    externalCallback(pushNotificationMessage);
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart:85]:
```suggestion
    await externalCallback(pushNotificationMessage);
```

File: packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    } else {
+      _eventQueue.add(notification);
+    }
+
+    await _dispatchToExternalHandle(notification);
+    await Future.wait(
+      _onNotificationReceivedInBackgroundCallbacks.map(
+        (callback) async {
+          await callback(notification);
+        },
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart:106]:
```suggestion
        (callback) => callback(notification),
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+  [
+    PushNotificationsHostApi,
+    ServiceProviderClient,
+    AmplifySecureStorage,
+    PluginUtilities,
+    AmplifyPushNotificationsFlutterApi,
+  ],
+)
+void main() {
+  const tokenReceivedEventChannel = EventChannel(
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:34]: This should be shared with the impl

File: packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart [hunk 13]
```diff
@@ -49,15 +68,27 @@ class AmplifyPushNotifications extends PushNotificationsPluginInterface {
   AmplifyPushNotifications({
     required ServiceProviderClient serviceProviderClient,
     required Future<void> Function() backgroundProcessor,
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart:70]: Why is this required?

Comment (dnys1) [reply packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart:70]: This class should probably be `abstract`. I don't see when it would be instantiated directly ever.

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+
+      testWidgetsFlutterBinding.defaultBinaryMessenger.setMockMethodCallHandler(
+        MethodChannel(tokenReceivedEventChannel.name),
+        handler,
+      );
+    });
+    test('should configure correctly', () async {
+      log.clear();
+      when(mockPushNotificationsHostApi.getLaunchNotification()).thenAnswer(
+        (_) => Future(() => standardAndroidPushMessage),
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:127]: Update throughout

```suggestion
        (_) async => standardAndroidPushMessage,
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+        mockServiceProviderClient.recordNotificationEvent(
+          eventType: anyNamed('eventType'),
+          notification: anyNamed('notification'),
+        ),
+      ).called(1);
+    });
+
+    test('should throw exception when configuring if there is no appId present',
+        () async {
+      final config0 = AmplifyConfig.fromJson(
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:148]: It's okay to shadow variable names

```suggestion
      final config = AmplifyConfig.fromJson(
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+          authProviderRepo: authProviderRepo,
+          config: config0,
+        ),
+        throwsA(const TypeMatcher<PushNotificationException>()),
+      );
+    });
+
+    test('should fail configure when registering device is unsuccessfull',
+        () async {
+      log.clear();
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:162]: This should be in `setUp`

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 17]
```diff
@@ -0,0 +1,444 @@
...
+      testWidgetsFlutterBinding.defaultBinaryMessenger.setMockMethodCallHandler(
+        MethodChannel(tokenReceivedEventChannel.name),
+        throwErrorHandler,
+      );
+      expect(
+        () async => plugin.configure(
+          authProviderRepo: authProviderRepo,
+          config: config,
+        ),
+        throwsA(const TypeMatcher<PushNotificationException>()),
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:172]: Should be checking more than just that it throws, e.g.

```suggestion
        throwsA(
          isA<PushNotificationException>()
            .having((e) => e.message, 'message', contains('token'),
        ),
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 18]
```diff
@@ -0,0 +1,444 @@
...
+    AmplifyPushNotificationsFlutterApi,
+  ],
+)
+void main() {
+  const tokenReceivedEventChannel = EventChannel(
+    'com.amazonaws.amplify/push_notification/event/TOKEN_RECEIVED',
+  );
+
+  final log = <MethodCall>[];
+  late TestWidgetsFlutterBinding testWidgetsFlutterBinding;
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:39]: Should only be called once per file.

```suggestion
  final testWidgetsFlutterBinding = TestWidgetsFlutterBinding.ensureInitialized();
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 19]
```diff
@@ -0,0 +1,444 @@
...
+      log.clear();
+      testWidgetsFlutterBinding.defaultBinaryMessenger.setMockMethodCallHandler(
+        MethodChannel(tokenReceivedEventChannel.name),
+        throwErrorHandler,
+      );
+      expect(
+        () async => plugin.configure(
+          authProviderRepo: authProviderRepo,
+          config: config,
+        ),
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:171]:
```suggestion
        plugin.configure(
          authProviderRepo: authProviderRepo,
          config: config,
        ),
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 20]
```diff
@@ -0,0 +1,444 @@
...
+      );
+      expect(
+        () async => plugin.launchNotification,
+        throwsA(const TypeMatcher<PushNotificationException>()),
+      );
+      expect(
+        () async => plugin.identifyUser(
+          userId: 'userId',
+          userProfile: AnalyticsUserProfile(),
+        ),
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:197]:
```suggestion
        plugin.identifyUser(
          userId: 'userId',
          userProfile: AnalyticsUserProfile(),
        ),
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 21]
```diff
@@ -0,0 +1,444 @@
...
+        throwsA(const TypeMatcher<PushNotificationException>()),
+      );
+      expect(
+        () async => plugin.requestPermissions(),
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:201]:
```suggestion
        plugin.requestPermissions,
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 22]
```diff
@@ -0,0 +1,444 @@
...
+        throwsA(const TypeMatcher<PushNotificationException>()),
+      );
+      expect(
+        () async => plugin.getPermissionStatus(),
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:205]:
```suggestion
        () async => plugin.getPermissionStatus,
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 23]
```diff
@@ -0,0 +1,444 @@
...
+      );
+
+      expect(
+        () async => plugin.setBadgeCount(42),
+        throwsA(const TypeMatcher<PushNotificationException>()),
+      );
+      expect(
+        () async => plugin.getBadgeCount(),
+        throwsA(const TypeMatcher<PushNotificationException>()),
+      );
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:216]: These two assertion actually caught the exception thrown for `!_iosCheck` I believe, which made the test inaccurate...

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 24]
```diff
@@ -0,0 +1,444 @@
...
+      expect(
+        res,
+        42,
+      );
+    });
+
+    test('setBadgeCount calls the native layer to set', () async {
+      await plugin.setBadgeCount(42);
+      verify(
+        mockPushNotificationsHostApi.setBadgeCount(any),
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:301]:
```suggestion
        mockPushNotificationsHostApi.setBadgeCount(42),
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 25]
```diff
@@ -0,0 +1,444 @@
...
+          final pref = await SharedPreferences.getInstance();
+          plugin = AmplifyPushNotifications(
+            serviceProviderClient: mockServiceProviderClient,
+            backgroundProcessor: () async => {},
+          )..onNotificationReceivedInBackground(
+              testGlobalCallbackFunction,
+            );
+          await Future.delayed(const Duration(seconds: 3), () {});
+
+          expect(pref.containsKey(externalHandleKey), true);
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:386]: nit: use a matcher when possible (same to other similar spots)
```suggestion
          expect(pref.containsKey(externalHandleKey), isTrue);
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 26]
```diff
@@ -0,0 +1,444 @@
...
+          SharedPreferences.setMockInitialValues({});
+          final pref = await SharedPreferences.getInstance();
+          plugin = AmplifyPushNotifications(
+            serviceProviderClient: mockServiceProviderClient,
+            backgroundProcessor: () async => {},
+            dependencyManager: dependencyManager,
+          )..onNotificationReceivedInBackground(
+              testGlobalCallbackFunction,
+            );
+          await Future.delayed(const Duration(seconds: 3), () {});
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:407]: nit: Does it really need 3 secs?

```suggestion
          await Future<void>.delayed(const Duration(seconds: 3));
```

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart [hunk 27]
```diff
@@ -0,0 +1,444 @@
...
+            serviceProviderClient: mockServiceProviderClient,
+            backgroundProcessor: () async => {},
+            dependencyManager: dependencyManager,
+          )..onNotificationReceivedInBackground(
+              (testGlobalCallbackFunction) {},
+            );
+          verify(
+            mockPushNotificationsNativeToFlutterApi
+                .registerOnReceivedInBackgroundCallback(
+              any,
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:434]: verify the `callback` is the one.

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart (hunk 26 above)
Comment (Samaritan1011001) [reply packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:407]: No, it's just some arbitrary time I put in to check. Can be lesser.

File: packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart (hunk 23 above)
Comment (Samaritan1011001) [reply packages/notifications/push/amplify_push_notifications/test/amplify_push_noitfications_impl_test.dart:216]: The ios check only logs an error and doesn't throw

File: packages/notifications/push/amplify_push_notifications_pinpoint/lib/src/pinpoint_provider.dart [hunk 28]
```diff
@@ -157,7 +168,7 @@ class PinpointProvider implements ServiceProviderClient {
       }
       if (data.containsKey(_androidCampaignTreatmentIdKey)) {
         campaign['treatment_id'] =
-            data[_androidCampaignTreatmentIdKey] as String;
+            data[_androidCampaignTreatmentIdKey].toString();
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/notifications/push/amplify_push_notifications_pinpoint/lib/src/pinpoint_provider.dart:171]: Unintentional. Reverting.

File: packages/amplify_core/lib/src/types/notifications/push/push_notification_message.dart (hunk 1 above)
Comment (Samaritan1011001) [reply packages/amplify_core/lib/src/types/notifications/push/push_notification_message.dart:76]: I added it to gracefully fail if the type of any field changes. Since we pass the rawData and other fields as we get in the native code, the type of them may be changed by cx when composing their messages I think. So, catching it and printing it rather than crashing. 

What would be the alternative here?

File: packages/notifications/push/amplify_push_notifications/test/push_notification_message_test.dart [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+      expect(parsedAndroidMessage.imageUrl, 'TEST_URL');
+
+      // ios shares the same key to pass along deeplink and url
+      final parseiOSMessage =
+          PushNotificationMessage.fromJson(imageUrliOSMessage);
+      expect(parseiOSMessage.imageUrl, 'TEST_URL');
+    });
+
+    test('should not crash when there is a type mismatch', () {
+      PushNotificationMessage.fromJson(defectivePushMessage);
```
Comment (HuiSF) [inline packages/notifications/push/amplify_push_notifications/test/push_notification_message_test.dart:50]:
```suggestion
      expect(
        () {
          PushNotificationMessage.fromJson(defectivePushMessage);
        },
        returnsNormally,
      )
```

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+      const OperatingSystem('android', ''),
+      () async {
+        SharedPreferences.setMockInitialValues({});
+        final pref = await SharedPreferences.getInstance();
+        AmplifyPushNotifications(
+          serviceProviderClient: MockServiceProviderClient(),
+          backgroundProcessor: () async => {},
+          // dependencyManager: dependencyManager,
+        ).onNotificationReceivedInBackground(
+          testGlobalCallbackFunction,
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:78]:
```suggestion
          expectAsync1(testGlobalCallbackFunction),
```

You need to set an expectation that this will be called. As it stands, this test would complete successfully whether the function is called or not.

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+  test('should call the external callback function', () async {
+    SharedPreferences.setMockInitialValues({});
+    void externalCallback(PushNotificationMessage pushMessage) {
+      expect(
+        pushMessage.title,
+        PushNotificationMessage.fromJson(standardAndroidPushMessage).title,
+      );
+    }
+
+    flutterApi.registerOnReceivedInBackgroundCallback(externalCallback);
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:61]:
```suggestion
    flutterApi.registerOnReceivedInBackgroundCallback(
      expectAsync1(externalCallback),
    );
```

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+void main() {
+  late AmplifyPushNotificationsFlutterApi flutterApi;
+
+  setUp(() {
+    TestWidgetsFlutterBinding.ensureInitialized();
+    flutterApi = AmplifyPushNotificationsFlutterApi.instance;
+  });
+
+  test('should queue events when the serviceClient is not ready', () {
+    SharedPreferences.setMockInitialValues({});
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:28]: Put this in `setUp`

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+        final pref = await SharedPreferences.getInstance();
+        AmplifyPushNotifications(
+          serviceProviderClient: MockServiceProviderClient(),
+          backgroundProcessor: () async => {},
+          // dependencyManager: dependencyManager,
+        ).onNotificationReceivedInBackground(
+          testGlobalCallbackFunction,
+        );
+        flutterApi
+            .registerOnReceivedInBackgroundCallback(testGlobalCallbackFunction);
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:81]: Why does this need to be called if `onNotificationReceived` is called above?

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart (hunk 30 above)
Comment (Samaritan1011001) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:78]: Tried `expectAsync1` but won't work for this case because `onNotificationReceivedInBackground` expects the callback function provided to be global or static but when I use `expectAsync1`, I believe it wraps it and makes it a local function?

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart [hunk 34]
```diff
@@ -0,0 +1,92 @@
...
+
+        await Future.delayed(const Duration(seconds: 3), () {});
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:83]: It's just an arbitrary time I set to wait for the callback function to be registered. Since we want to keep the `onNotificationReceivedInBackground` async we are not waiting for the registration process to finish, hence I wait before triggering the function that is responsible to handle the notification coming from the native side.

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart (hunk 31 above)
Comment (Samaritan1011001) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:61]: Since I changed `flutterApi` to be a `singleton`, any registered callback here will be registered in the next tests as well. Making this `expectAsync1` errors out since the next test also invokes `externalCallback`.

File: packages/notifications/push/amplify_push_notifications/test/push_notification_message_test.dart (hunk 29 above)
Comment (Samaritan1011001) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_message_test.dart:50]: Thanks, but removing this test based on the decision to remove the try catch block in `fromJson`.

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart (hunk 34 above)
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:83]: If it's arbitrary please use a smaller delay to make the tests run faster. The tests may be roughly 6 seconds slower waiting for this arbitrary time :D

**Review by HuiSF — APPROVED** [review]

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart (hunk 30 above)
Comment (dnys1) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:80]: There should be no state left over between tests - this will greatly increase the risk of false positives in tests (like here potentially). I would recommend adding a `reset` method to the static class and call that in between tests.

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart (hunk 31 above)
Comment (dnys1) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart]: There should be no state left over between tests - this will greatly increase the risk of false positives in tests (like here potentially). I would recommend adding a reset method to the static class and call that in between tests.

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart (hunk 30 above)
Comment (dnys1) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:80]: You can implement it yourself with completers

File: packages/notifications/push/amplify_push_notifications/android/build.gradle [hunk 35]
```diff
@@ -71,6 +66,16 @@ def amplifyFlutter = project.rootProject.findProject(':amplify_flutter_android')
...
-    implementation 'com.amplifyframework:aws-push-notifications-pinpoint-utils:2.1.0'
+    // Import support library for Amplify push utils
+    implementation 'com.amplifyframework:aws-push-notifications-pinpoint-common:2.5.0'
+    implementation "org.jetbrains.kotlinx:kotlinx-serialization-json:1.4.1"
     implementation project(path: ':flutter_plugin_android_lifecycle')
+    implementation 'androidx.test:core-ktx:1.5.0'
+
+    testImplementation 'junit:junit:4.13.2'
+    testImplementation 'io.mockk:mockk:1.13.4'
+    testImplementation 'org.slf4j:slf4j-nop:2.0.7'
```
[the remark this answers is not in the document]
Comment (dnys1) [reply packages/notifications/push/amplify_push_notifications/android/build.gradle:77]: I'm guessing it's used by `mockk`, then? None of the packages we typically use require this. Can you add a note that it's needed for `mockk`?

File: packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart (hunk 13 above)
Comment (dnys1) [reply packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart:63]: Can you update to make this abstract?

File: packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart [hunk 36]
```diff
@@ -0,0 +1,129 @@
+import 'dart:async';
```
Comment (dnys1) [inline packages/notifications/push/amplify_push_notifications/lib/src/push_notifications_flutter_api.dart:1]: Still missing several license headers throughout. To add them all at once:

```
cd packages/notifications
../../tool/license.sh
```

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart (hunk 34 above)
[the remark this answers is not in the document]
Comment (dnys1) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:83]: If you implement the `expectAsync` pattern with a completer, you can avoid this wait

**Review by dnys1 — APPROVED** [review]

File: packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart (hunk 34 above)
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/notifications/push/amplify_push_notifications/test/push_notification_flutter_api_test.dart:83]: I am having to wait in-between callback registration and notification arrival because when registering we are using `unawaited` within the `onNotificationReceivedInBackground` [function](https://github.com/aws-amplify/amplify-flutter/blob/feat/push/android-utils/packages/notifications/push/amplify_push_notifications/lib/src/amplify_push_notifications_impl.dart#L162-L167). This choice was made to allow for using the API without awaiting it but as we can see it can introduce problems if the notification arrives too early as mimicked in our test by the use of `flutterApi.onNotificationReceivedInBackground`. 

I don't completely understand how completer is useful here but one proposal I have is to make the `onNotificationReceivedInBackground` API function async which will ensure the async operation to store the callback handle in shared preference is successful before it's ready to receive notifications.
